### PR TITLE
Fixed wrong synchronization of threads in throttling counting

### PR DESCRIPTION
**Pull request by openhubframework**

### CODE REVIEW

**Review by hanusto — CHANGES_REQUESTED** [review]
From my side, it is necessary to do:

1) read contribution notes - https://openhubframework.atlassian.net/wiki/display/OHF/Contribution
2) read contribution rules and principles - https://openhubframework.atlassian.net/wiki/display/OHF/Rules+and+principles
3) for all java based files please change version to 2.0.0 (next release version will be 2.0.0)
4) please avoid to use full author tag (do not add e-mail address)
5) it is necessary to rebase these commits on to develop and solve some conflicts
6) please improve title and description of this PR
7) it is possible to squash all commits together?

After I can continue in review.

File: common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationBlock.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.openhubframework.openhub.common.synchronization;
+
+/**
+ * Method {@link #doInSynchronizationPart()} in this interface can be synchronized by one one value in
```
Comment (pjuza) [inline common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationBlock.java]: mistake "one one"

File: common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+import org.springframework.util.Assert;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Executor for {@link SynchronizationBlock} which synchronized more then one threads by one value.
```
Comment (pjuza) [inline common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java]: which -> that

File: common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+ * <p>
+ * For synchronization by one value call {@link #execute(SynchronizationBlock, String, Object)}, where in method
+ * {@link SynchronizationBlock#doInSynchronizationPart()} are code which will be by synchronized by value.
```
Comment (pjuza) [inline common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java]: which -> that + will be synchronized by value

**Review by pjuza — COMMENTED** [review]
Please add summary of what is going on in this task (https://openhubframework.atlassian.net/browse/OHFJIRA-14) - we know problem and it's necessary to add task description, what is implemented

File: common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationNoResultBlock.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    public final <T> T doInSynchronizationPart() {
+        doInSynchronizationPartNoResult();
+        return null;
+    }
+
+    /**
+     * Method will be synchronized by one value calling
+     * {@link SynchronizationExecutor#execute(SynchronizationBlock, String, Object)}.
+     */
+    protected abstract void doInSynchronizationPartNoResult();
```
Comment (pjuza) [inline common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationNoResultBlock.java]: When the class has name SynchronizationNoResultBlock then method could be shorter, for example doInSynchro or synchroBlock ...

File: common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+ * </p>
```
Comment (pjuza) [inline common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java]: Little bit more <p> :)

**Comment by hanusto** [comment]
Maybe we should consider new PR in Camel Core:

- https://github.com/apache/camel/pull/1404
- https://github.com/apache/camel/pull/1400 ([ThrottlingExceptionRoutePolicy](https://github.com/apache/camel/pull/1400/files#diff-033333669cf0e63ff86d4fe88d2eae7fR52))

File: common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    //--------------------------------------------- STATIC -------------------------------------------------------------
+
+    /**
+     * Gets instance of this class.
+     *
+     * @return instance of this class
+     */
+    public synchronized static SynchronizationExecutor getInstance() {
```
Comment (hanusto) [inline common/src/main/java/org/openhubframework/openhub/common/synchronization/SynchronizationExecutor.java:198]: Synchronization of lazy loaded singleton is broken. Please you static eager initialization - http://www.javaworld.com/article/2074979/java-concurrency/double-checked-locking--clever--but-broken.html?page=2.

File: core/src/main/java/org/openhubframework/openhub/core/throttling/ThrottleCounterMemoryImpl.java [hunk 7]
```diff
@@ -45,81 +43,60 @@
 
     private static final int DUMP_PERIOD = 60;
 
+    private static final String SYNC_THROTTLING_TYPE = "SYNCHRONIZATION_THROTTLING";
```
Comment (hanusto) [inline core/src/main/java/org/openhubframework/openhub/core/throttling/ThrottleCounterMemoryImpl.java:51]: Maybe `THROTTLING_SYNCHRONIZATION` is better name.

File: core/src/main/java/org/openhubframework/openhub/core/throttling/ThrottleCounterMemoryImpl.java [hunk 8]
```diff
@@ -45,81 +43,60 @@
...
-            if (requests.get(throttleScope) == null) {
-                requests.put(throttleScope, new Stack<Long>());
-            }
+                if (requests.get(throttleScope) == null) {
+                    requests.put(throttleScope, new Stack<Long>());
+                }
 
-            long now = DateTime.now().getMillis();
-            long from = now - (interval * 1000);
+                long now = DateTime.now().getMillis();
```
Comment (hanusto) [inline core/src/main/java/org/openhubframework/openhub/core/throttling/ThrottleCounterMemoryImpl.java:76]: We use spaces for indent, no tabs - see https://openhubframework.atlassian.net/wiki/display/OHF/Source+code+conventions#Sourcecodeconventions-Syntaxrules.

**Review by hanusto — CHANGES_REQUESTED** [review]
Small changes are necessary (reformat code) and consider if lazy loaded singleton is necessary. After clarification it can be approved (merged).

**Review by hanusto — APPROVED** [review]

**Pull request merged**